tanner/utils/aiodocker_helper.py

```diff
@@ -0,0 +1,82 @@
+import aiodocker
+import logging
+
+from tanner.config import TannerConfig
+
+
+class AIODockerHelper:
+    def __init__(self):
+
+        self.logger = logging.getLogger('tanner.aiodocker_helper.AIODockerHelper')
+
+        self.docker_client = aiodocker.Docker()
+        self.host_image = TannerConfig.get('DOCKER', 'host_image')
+
+    async def setup_host_image(self, remote_path=None, tag=None):
+
+        try:
+            if remote_path and tag is not None:
+                params = {"tag": tag, "remote": remote_path}
+                await self.docker_client.images.build(**params)
+
+            image = await self.docker_client.images.list(filter=self.host_image)
+            if not image:
+                await self.docker_client.images.pull(self.host_image)
+
+        except aiodocker.exceptions.DockerError as docker_error:
+            self.logger.exception('Error while pulling %s image %s', self.host_image, docker_error)
+
+    async def get_container(self, container_name):
+        container = None
+        try:
+            container = await self.docker_client.containers.get(container=container_name)
+
+        except aiodocker.exceptions.DockerError as server_error:
+            self.logger.exception('Error while fetching %s container %s', container_name, server_error)
+        return container
+
+    async def create_container(self, container_name, cmd=None, image=None):
+        await self.setup_host_image()
+        container = None
+        if image is None:
+            image = self.host_image
+
+        config = {
+            "Cmd": cmd,
+            "Image": image,
+        }
+        try:
+            container = await self.docker_client.containers.create_or_replace(config=config, name=container_name)
+
+        except (aiodocker.exceptions.DockerError or aiodocker.exceptions.DockerContainerError) as docker_error:
+            self.logger.exception('Error while creating a container %s', docker_error)
+        return container
+
+    async def execute_cmd(self, cmd, image=None):
+        execute_result = None
+        try:
+            if image is None:
+                image = self.host_image
+
+            config = {"Cmd": cmd, "Image": image}
+            container = await self.docker_client.containers.run(config=config)
```

Comment:
As far as I understood, you create ll the time new containers with a random name, how you suppose to clean it?

Reply:
For now these containers have to be deleted manually because of the risk other containers can also be deleted if we try to delete them all. Do you have something in mind?

Reply:
Previously we had a procedure to delete containers in session routine, the idea is to have an associated container with session and delete it by name while deleting the session

Reply:
Ahh! Yes, then in this case we should use create method to run commands in emulators right?

Reply:
You can use `create`, `get` and `start` methods
or
You can still create the containers with `run` and delete it right after the execution.

Reply:
I don't know what is better in terms of architecture and performance, you can test it :)

Reply:
I think the second method works best.

Reply:
The only problem I can imagine with a second approach is a (possible) inconsistency in the sequence of requests. And also, have you measured what is faster: always create and delete new one or get existing and start it? I would be happy to see such compariosn

Reply:
So, I analyzed the performance of both methods - (averaged multiple runs)

1. **create_or_replace:** gets the container and replaces new config or creates a new one + starting the container - **1.4206 sec**  (deleting time and logs output time has to be added)

2. **run + delete:** creates a new container and deleting [ had to force :( ] container + checking output using logs- **1.5324 sec**

Reply:
I think it's not a significant performance difference, but the method you will prefer would be final :)